File: app/src/main/java/dev/yiray/qickmvivm/ui/home/TaskFragment.java
package dev.yiray.qickmvivm.ui.home;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import dev.yiray.qickmvivm.base.BaseMVIVMFragment;
import dev.yiray.qickmvivm.databinding.FragmentTaskBinding;
import dev.yiray.qickmvivm.databinding.ItemTaskBinding;
import dev.yiray.qickmvivm.model.Status;
import dev.yiray.qickmvivm.model.Task;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.subjects.PublishSubject;

public class TaskFragment extends BaseMVIVMFragment<HomeView.Action, HomeViewState, HomeViewModel>
        implements HomeView.Action
{
    private FragmentTaskBinding binding;
    private List<Task> taskList;
    private PublishSubject<HomeView.UpdateTask> mUpdateTaskSubject;
    private TaskRecyclerViewAdapter adapter;

    public TaskFragment() {
    }

    @SuppressWarnings("unused")
    public static TaskFragment newInstance(int columnCount) {
        return new TaskFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mUpdateTaskSubject = PublishSubject.create();
        taskList = new ArrayList<>();
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentTaskBinding.inflate(inflater, container, false);
        adapter = new TaskRecyclerViewAdapter();

        RecyclerView recyclerView = binding.list;
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);

        binding.btnAdd.setOnClickListener(this::onAdd);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mViewModel.bindIntent(this);
    }

    @Override
    protected HomeViewModel bindViewModel() {
        return new ViewModelProvider(requireActivity()).get(HomeViewModel.class);
    }

    @Override
    public void render(HomeViewState state) {
        List<Task> updatedList = state.toViewState().getTaskList();

        if (updatedList != null) {
            taskList = state.toViewState().getTaskList();
        }
        adapter.notifyDataSetChanged();
    }

    @Override
    public void handleSideEffect(HomeViewState state, HomeViewState sideEffect) {

    }

    @Override
    public void displayErrorMsg(String message) {

    }

    @Override
    public Observable<CharSequence> observableInput() {
        return Observable.empty();
    }

    @Override
    public Observable<Boolean> observableCheckBox() {
        return Observable.empty();
    }

    @Override
    public Observable<Boolean> observableNextPage() {
        return Observable.empty();
    }

    @Override
    public Observable<HomeView.UpdateTask> observableUpdated() {
        return mUpdateTaskSubject;
    }

    private void onAdd(View v) {
        EditText editText = binding.etTask;

        if (!TextUtils.isEmpty(editText.getText().toString())) {
            mUpdateTaskSubject.onNext(
                    new HomeView.UpdateTask(taskList.size(), new Task(editText.getText().toString()))
            );
            editText.setText("");
        }
    }

    public class TaskRecyclerViewAdapter extends RecyclerView.Adapter<TaskRecyclerViewAdapter.ViewHolder> {

        @NotNull
        @Override
        public TaskRecyclerViewAdapter.ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
            ItemTaskBinding binding = ItemTaskBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new TaskRecyclerViewAdapter.ViewHolder(binding);
        }

        @Override
        public void onBindViewHolder(final TaskRecyclerViewAdapter.ViewHolder holder, int position) {
            holder.mTask = taskList.get(position);
            holder.tvName.setText(holder.mTask.getName());

            String status = "";

            switch (holder.mTask.getStatus()) {
                case OPEN:
                    status = "☐";
                    break;
                case COMPLETE:
                    status = "☑";
                    break;
            }

            holder.tvStatus.setText(status);
            holder.itemView.setOnClickListener(v -> mUpdateTaskSubject.onNext(
                    new HomeView.UpdateTask(position, new Task(holder.mTask.getName(), Status.COMPLETE))
            ));
        }

        @Override
        public int getItemCount() {
            return taskList.size();
        }

        public class ViewHolder extends RecyclerView.ViewHolder {
            public final View itemView;
            public final TextView tvName;
            public final TextView tvStatus;

            public Task mTask;

            public ViewHolder(ItemTaskBinding binding) {
                super(binding.getRoot());
                itemView = binding.getRoot();
                tvName = binding.name;
                tvStatus = binding.status;
            }

            @NotNull
            @Override
            public String toString() {
                return super.toString() + " '" + tvName.getText() + "'";
            }
        }
    }
}

File: app/src/main/java/dev/yiray/livedatamvi/ui/home/TaskFragment.kt
package dev.yiray.livedatamvi.ui.home

import dev.yiray.livedatamvi.ui.home.HomeView.UpdateTask
import dev.yiray.livedatamvi.base.BaseMVIVMFragment
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.LinearLayoutManager
import android.text.TextUtils
import android.view.View
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import dev.yiray.livedatamvi.databinding.FragmentTaskBinding
import dev.yiray.livedatamvi.databinding.ItemTaskBinding
import dev.yiray.livedatamvi.model.Status
import dev.yiray.livedatamvi.model.Task
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.ArrayList

class TaskFragment : BaseMVIVMFragment<HomeView.Action, HomeViewState, HomeViewModel>(),
    HomeView.Action {
    private var binding: FragmentTaskBinding? = null
    private var taskList: List<Task?>? = null
    private var mUpdateTaskSubject = MutableSharedFlow<UpdateTask>()
    private var adapter: TaskRecyclerViewAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        taskList = ArrayList()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentTaskBinding.inflate(inflater, container, false)

        adapter = TaskRecyclerViewAdapter()
        val recyclerView = binding!!.list
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = adapter
        binding!!.btnAdd.setOnClickListener { v: View -> onAdd() }

        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mViewModel!!.bindIntent(this)
    }

    override fun bindViewModel(): HomeViewModel {
        return ViewModelProvider(requireActivity()).get(HomeViewModel::class.java)
    }

    override fun render(state: HomeViewState) {
        val nowSize = taskList!!.size
        val updatedList = state.toViewState()!!.taskList
        if (updatedList != null) {
            taskList = state.toViewState()!!.taskList
        }
//        adapter!!.notifyDataSetChanged()
        state.toViewState()!!.lastModifiedPosition?.let {
//            if (nowSize < taskList!!.size) {
//                adapter!!.notifyItemRangeChanged(0, taskList!!.size)
//                return
//            }
            adapter!!.notifyItemChanged(it)
        }
    }

    override fun handleSideEffect(state: HomeViewState, sideEffect: HomeViewState) {}
    override fun displayErrorMsg(message: String?) {}
    override fun observableInput(): Flow<CharSequence> {
        return flow { }
    }

    override fun observableCheckBox(): Flow<Boolean> {
        return flow { }
    }

    override fun observableNextPage(): Flow<Boolean> {
        return flow { }
    }

    override fun observableUpdated(): Flow<UpdateTask>? {
        return mUpdateTaskSubject
    }

    private fun onAdd() {
        val editText = binding!!.etTask
        if (!TextUtils.isEmpty(editText.text.toString())) {
            lifecycleScope.launch {
                mUpdateTaskSubject.emit(
                    UpdateTask(taskList!!.size, Task(editText.text.toString()))
                )
            }
            editText.setText("")
        }
    }

    inner class TaskRecyclerViewAdapter :
        RecyclerView.Adapter<TaskRecyclerViewAdapter.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val binding =
                ItemTaskBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return ViewHolder(binding)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.mTask = taskList!![position]
            holder.tvName.text = holder.mTask!!.name
            var status = ""
            status = when (holder.mTask!!.status) {
                Status.OPEN -> "☐"
                Status.COMPLETE -> "☑"
            }
            holder.tvStatus.text = status
            holder.itemView.setOnClickListener {
                lifecycleScope.launch {
                    mUpdateTaskSubject.emit(
                        UpdateTask(position, Task(holder.mTask!!.name, Status.COMPLETE))
                    )
                }
            }
        }

        override fun getItemCount(): Int {
            return taskList!!.size
        }

        inner class ViewHolder(binding: ItemTaskBinding) : RecyclerView.ViewHolder(binding.root) {
            val tvName: TextView = binding.name
            val tvStatus: TextView = binding.status
            var mTask: Task? = null
            override fun toString(): String {
                return super.toString() + " '" + tvName.text + "'"
            }

        }
    }

    companion object {
        fun newInstance(columnCount: Int): TaskFragment {
            return TaskFragment()
        }
    }
}